### PR TITLE
Condor rc3

### LINKED ISSUE #137: CEP18 / Migration to Condor 2.0 - Hash Types Update
Migration to Condor 2.0 - Hash Types Update 

#PR to merge post Condor release

### PR DESCRIPTION
**Pull request by deuszex**
2.0.0-rc3 compatible.
Migration entrypoints for balance and allowance keys to move from old Key formats to new ones (AddressableEntity).
Events:
EventsMode::CES events included (forked original to update to rc3).
EventsMode::Native events included (Casper Native events). 
EventsMode::NativeNCES (uses both Native and CES eventing).
Compiles on latest rust.

Added 1.5.6 fixture(state snapshot) to test migration from 1.5.6 to 2.0.0-rc3.

TODO: Fixture guide/tutorial.

Added entrypoint:
- `migrate_user_allowance_keys`: 
- `migrate_user_balance_keys`
- `migrate_user_sec_keys`
The above entrypoints change the stored keys from the old style (Key::Hash/Key::Account) to the new one (Key::AddressableEntity) which will be used in the contract going forward. The entrypoints also take `events` and `revert` boolean arguments. Supplying `false` to the `events` argument turns the eventing off during the current calling of the entrypoint, since the size of the events can be rather large depending on number of arguments provided. Supplying `true` as the `revert` argument on the other hand will cause the call to immediately revert upon meeting an error, which may help in determining the point of failure if encountered one.
- `change_events_mode` (needs testing): to change the events mode without needing to upgrade the whole contract.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,14 +1,13 @@
-PINNED_TOOLCHAIN := $(shell cat rust-toolchain)
-
 prepare:
 	rustup target add wasm32-unknown-unknown
 	rustup component add clippy --toolchain ${PINNED_TOOLCHAIN}
 	rustup component add rustfmt --toolchain ${PINNED_TOOLCHAIN}
+	rustup component add rust-src --toolchain ${PINNED_TOOLCHAIN}
```
Comment (gRoussac) [inline Makefile]: `PINNED_TOOLCHAIN` seems to be missing

```
PINNED_TOOLCHAIN := $(shell cat rust-toolchain)
```

or

```
PINNED_TOOLCHAIN := $(shell cat cep18/rust-toolchain)
```

File: cep18/src/constants.rs [hunk 2]
```diff
@@ -67,3 +74,4 @@ pub const BURNER_LIST: &str = "burner_list";
 pub const NONE_LIST: &str = "none_list";
 pub const MINT_AND_BURN_LIST: &str = "mint_and_burn_list";
```
Comment (gRoussac) [inline cep18/src/constants.rs]: MINT_AND_BURN_LIST and BURNER_LIST seem unsued

File: cep18/src/main.rs [hunk 3]
```diff
@@ -351,38 +407,387 @@ pub extern "C" fn change_security() {
...
+    sec_check(vec![SecurityBadge::Admin]);
+    let events_mode: EventsMode = EventsMode::try_from(get_named_arg::<u8>(EVENTS_MODE))
+        .unwrap_or_revert_with(Cep18Error::InvalidEventsMode);
+
+    let events_mode_u8 = events_mode as u8;
+    put_key(EVENTS_MODE, storage::new_uref(events_mode_u8).into());
+
+    match events_mode {
+        EventsMode::NoEvents => {}
+        EventsMode::CES => init_events(),
```
Comment (gRoussac) [inline cep18/src/main.rs:725]: @deuszex As discussed please check this bug on cep-78, https://github.com/casper-ecosystem/cep-78-enhanced-nft/issues/261 

Please try a "migration when EventsMode::CES to EventsMode::CES", I would assume you need to check previous events_mode before calling init_events again.

https://github.com/casper-ecosystem/cep-78-enhanced-nft/pull/263/files

File: tests/src/utility/constants.rs [hunk 4]
```diff
@@ -67,10 +72,12 @@ pub const METHOD_TRANSFER_AS_STORED_CONTRACT: &str = "transfer_as_stored_contrac
 pub const METHOD_APPROVE_AS_STORED_CONTRACT: &str = "approve_as_stored_contract";
 pub const METHOD_FROM_AS_STORED_CONTRACT: &str = "transfer_from_as_stored_contract";
 
-pub const TOKEN_OWNER_ADDRESS_1: Key = Key::Account(AccountHash::new([42; 32]));
+pub const TOKEN_OWNER_ADDRESS_1: Key = Key::AddressableEntity(EntityAddr::Account([42; 32]));
 pub const TOKEN_OWNER_AMOUNT_1: u64 = 1_000_000;
-pub const TOKEN_OWNER_ADDRESS_2: Key = Key::Hash([42; 32]);
+pub const TOKEN_OWNER_ADDRESS_2: Key = Key::AddressableEntity(EntityAddr::SmartContract([42; 32]));
```
Comment (gRoussac) [inline tests/src/utility/constants.rs:60]: @deuszex I believe contracts should not be owners but only EntityAddr::Account and Package

Comment (deuszex) [reply tests/src/utility/constants.rs:60]: Correct, that was a mismanagement on my end

File: tests/src/utility/installer_request_builders.rs [hunk 5]
```diff
@@ -30,13 +30,11 @@ pub(crate) fn invert_cep18_address(address: Key) -> Key {
         Key::Hash(contract_hash) => Key::Account(AccountHash::new(contract_hash)),
         Key::AddressableEntity(entity_addr) => match entity_addr {
             EntityAddr::System(_) => panic!("Unsupported Key variant"),
-            EntityAddr::Account(account) => {
-                Key::AddressableEntity(EntityAddr::SmartContract(account))
-            }
-            EntityAddr::SmartContract(contract) => {
-                Key::AddressableEntity(EntityAddr::Account(contract))
-            }
+            EntityAddr::Account(account) =>
```
Comment (gRoussac) [inline tests/src/utility/installer_request_builders.rs]: @deuszex this seems strange Account to Package ?

Comment (gRoussac) [reply tests/src/utility/installer_request_builders.rs]: oh invert! I get it

**Comment by sczembor** [comment]
The current approach relies on a user-provided key map, which introduces the risk of errors and could be cumbersome for larger migrations. Also, it doesn't account for potential "race" if the contract is used post condor and before/while the migration process.

I'd like to propose an alternative "on-demand" migration. This approach would:

1. Trigger migration automatically whenever a account or contract (AddressableEntity) interacts with the smart contract.
2. Reconstruct the old key (AccountHash, ContractHash) and transfer the balance, allowances to the new `AddressableEntity` key structure if the migration didnt already occurred.

This eliminates the need for a manual key map and reduces the risk of errors. While it might add some complexity, I believe it offers a more robust and scalable solution.

I'm happy to discuss it and see your point of view. @deuszex @gRoussac

**Comment by deuszex** [comment]
@sczembor That is the so collad "lazy migration" that we have discussed already on our meetings on which you are also present, and as I've said on the mentioned meetings, these decision will go through the proper chain of command, when the relevant personnel are available.

**Comment by deuszex** [comment]
Closing in favour of another PR

**Pull request closed without merging**